pkg/keycloak/client.go

```diff
@@ -48,50 +50,151 @@ type TokenResponse struct {
+	jsonValue, _ := json.Marshal(realm)
+
+	logrus.Debugf("creating realm, %v, %v", realm, string(jsonValue))
+
+	req, err := http.NewRequest(
+		"POST",
+		fmt.Sprintf("%s/auth/admin/realms/", c.URL),
+		bytes.NewBuffer(jsonValue),
+	)
+	if err != nil {
+		logrus.Errorf("error creating request %+v", err)
+		return err
+	}
+	req.Header.Set("Content-Type", "application/json")
+	req.Header.Add("Authorization", "Bearer " + c.token)
+	res, err := c.Requester.Do(req)
+
+	if err != nil {
+		logrus.Errorf("error on request %+v", err)
+		return err
+	}
+
+	logrus.Debugf("response status: %v, %v", res.StatusCode, res.Status)
+	if res.StatusCode != 201 {
+		return errors.New("failed to create realm: " + " (" + strconv.Itoa(res.StatusCode) + ") " + res.Status)
+	}
 
-	return map[string]*v1alpha1.KeycloakRealm{}, nil
+	logrus.Debugf("response:", res)
+
+	return nil
 }
 
 func (c *Client) GetRealm(name string) (*v1alpha1.KeycloakRealm, error) {
-	return &v1alpha1.KeycloakRealm{}, nil
+	req, err := http.NewRequest(
+		"GET",
+		fmt.Sprintf("%s/auth/admin/realms/%s", c.URL, name),
+		nil,
+	)
+	if err != nil {
+		logrus.Errorf("error creating request %+v", err)
+		return nil, err
+	}
+	req.Header.Add("Authorization", "Bearer " + c.token)
+	res, err := c.Requester.Do(req)
+
+	if err != nil {
+		logrus.Errorf("error on request %+v", err)
+		return nil, err
+	}
+
+	logrus.Debugf("response status: %v, %v", res.StatusCode, res.Status)
+	if res.StatusCode < 200 || res.StatusCode > 299 {
+		return nil, errors.New("failed to update realm: " + " (" + strconv.Itoa(res.StatusCode) + ") " + res.Status)
+	}
+
+	body, err := ioutil.ReadAll(res.Body)
+	if err != nil {
+		logrus.Errorf("error reading response %+v", err)
+		return nil, errors.Wrap(err, "error reading realms list response")
+	}
+
+	logrus.Debugf("realm: %+v\n", string(body))
+
+	var realm *v1alpha1.KeycloakRealm
+	err = json.Unmarshal(body, &realm)
+
+	if err != nil {
+		logrus.Error(err)
+	}
+	logrus.Debugf("realm = %#v", realm)
+
+
+	return realm, nil
 }
 
 func (c *Client) UpdateRealm(realm *v1alpha1.KeycloakRealm) error {
-	return nil
-}
+	jsonValue, _ := json.Marshal(realm)
+
+	logrus.Debugf("updating realm, %v, %v", realm, string(jsonValue))
+
+	req, err := http.NewRequest(
+		"PUT",
+		fmt.Sprintf("%s/auth/admin/realms/%s", c.URL, realm.ID),
+		bytes.NewBuffer(jsonValue),
+	)
+	if err != nil {
+		logrus.Errorf("error creating request %+v", err)
+		return err
+	}
+	req.Header.Set("Content-Type", "application/json")
+	req.Header.Add("Authorization", "Bearer " + c.token)
+	res, err := c.Requester.Do(req)
+
+	if err != nil {
+		logrus.Errorf("error on request %+v", err)
+		return err
+	}
+
+	logrus.Debugf("response status: %v, %v", res.StatusCode, res.Status)
+	if res.StatusCode < 200 || res.StatusCode > 299 {
+		return errors.New("failed to update realm: " + " (" + strconv.Itoa(res.StatusCode) + ") " + res.Status)
+	}
+
+	logrus.Debugf("response:", res)
```

Comment:
Lots of duplication across all these methods. Should look at refactoring some of this, but can be done later once we are happy with the client functions we have.